spinalcordtoolbox/scripts/sct_dmri_display_bvecs.py

```diff
@@ -52,37 +68,93 @@ def get_parser():
 
+
+def create_custom_legend(fig, shell_colors, bvals):
+    """
+    Create single custom legend for whole figure
+    :param fig: figure the legend will be created for
+    :param shell_colors: dict with b-values and colors
+    :param bvals: ndarray with all b-values
+    :return:
+    """
+
+    # count number of bvals for individual shells
+    bvals_per_shell = pd.Series(bvals[bvals > bzero]).value_counts()
+
+    # Create single custom legend for whole figure with several subplots
+    lines = list()  # initialize list for individual symbols in the legend
+    labels = list()  # initialize list for individual text labels in the legend
+    # Loop across legend elements
+    for key, value in shell_colors.items():
+        lines.append(Line2D([0], [0], color=value, marker='o', markersize=8, alpha=0.7, linestyle=''))
+        labels.append('b-values = {} (n = {})'.format(str(int(key)), str(bvals_per_shell[key])))
+
+    plt.tight_layout()
+    # Insert legend below subplots, NB - this line has to be below the plt.tight_layout()
+    legend = fig.legend(lines, labels, loc='lower left', bbox_to_anchor=(0.1, 0),
+                        bbox_transform=plt.gcf().transFigure, ncol=len(lines), fontsize=10)
+    # Change box's frame color to black
+    frame = legend.get_frame()
+    frame.set_edgecolor('black')
+    # tight layout of whole figure and shift master title up
+    fig.tight_layout()
+    fig.subplots_adjust(top=0.88, bottom=0.1)
+
+
 # MAIN
 # ==========================================================================================
 
-
 def main(argv=None):
     parser = get_parser()
     arguments = parser.parse_args(argv)
     verbose = arguments.v
     set_global_loglevel(verbose=verbose)
 
     fname_bvecs = arguments.bvec
+    if arguments.bval is not None:
+        fname_bvals = arguments.bval
+        # Read bvals and bvecs files
+        bvals, bvecs = read_bvals_bvecs(fname_bvals, fname_bvecs)
+    else:
+        # Read only bvecs file
+        _, bvecs = read_bvals_bvecs(None, fname_bvecs)
 
-    # Read bvecs
-    bvecs = read_bvals_bvecs(fname_bvecs, None)
-    bvecs = bvecs[0]
     # if first dimension is not equal to 3 (x,y,z), transpose bvecs file
     if not bvecs.shape[0] == 3:
         bvecs = bvecs.transpose()
-    x, y, z = bvecs[0], bvecs[1], bvecs[2]
+
+    # bvals file was passed
+    if arguments.bval is not None:
+        # multiply unit b-vectors by b-values
+        x, y, z = bvecs[0] * bvals, bvecs[1] * bvals, bvecs[2] * bvals
+    # bvals file was not passed
+    else:
+        x, y, z = bvecs[0], bvecs[1], bvecs[2]
+        # create dummy unit bvals array (necessary fot scatter plots)
+        bvals = np.repeat(1, bvecs.shape[1])
+
+    # Assign scatter markers to unique bvals
+    shell_colors = {}
+    index = 0
+    for unique_bval in np.unique(bvals):
+        # skip b=0
+        if unique_bval < bzero:
+            continue
+        shell_colors[unique_bval] = colors[index]
+        index += 1
```

Comment:
Great point! I was considering the proper solution. Yes, there's theoretical possibility to have >6 shells in diffusion sequence. Then, it would fail. I'll check `cycler`, thanks!

Reply:
Hm, I haven't found satisfactory workaround with `cycler` library, it looks it can iterate only across predefined list of colors (don't solve our problem). But I haven't study it so deeply.
Instead I found quick workaround using `matplotlib.cm` and iterator, see commit https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3390/commits/a6886da7caa1efb329701d0493bde9c0005162fd (and inspiration [here](https://www.kite.com/python/answers/how-to-set-different-colors-for-each-series-in-a-matplotbib-scatter-plot-in-python))